---
title: "BXD and BXA volumes"
author: "Jason Lerch"
date: "19/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(ggplot2)
library(data.tree)
```

Load input data

```{r}
#File missing 
# BXDleafvolumes <- readRDS("./jason_example/BXDsymleafnodesvolumes.RDS")

BXAleafvolumes <- readRDS("./jason_example/BXAsymleafnodesvolumes.RDS")

class(BXAleafvolumes)
```
```{r}
dim(BXAleafvolumes)
```
```{r}
BXAleafvolumes[1:5,1:5]
```

Contains bilateral DSURQE labels.


```{r}
BXDmice <- readRDS("./jason_example/BXDmousedescriptors.RDS")
BXAmice <- readRDS("./jason_example/BXAmousedescriptors.RDS")

class(BXAmice)
```
```{r}
dim(BXAmice)
```
```{r}
head(BXAmice)
```

Information about mice.


```{r}
BXDtree <- readRDS("./jason_example/BXDdatatree.RDS")
BXAtree <- readRDS("./jason_example/BXAdatatree.RDS")
```

Merge BXD and BXA datasets

```{r}
# BXDmicesimple  <- BXDmice %>%
#   rename(Strain = strain) %>%
#   mutate(Cohort = "BXD") %>%
#   select(Strain, Cohort)
# 
# BXAmicesimple <- BXAmice %>%
#   mutate(Strain = fct_recode(Strain, B6 = "C57bl/6j"),
#          Cohort = "BXA") %>%
#   select(Strain, Cohort)
# 
# BXAandBXDmice <- rbind(BXAmicesimple, BXDmicesimple)
# BXAandBXDleafvolumes <- rbind(BXAleafvolumes, BXDleafvolumes)

BXAandBXDmice <- readRDS('./jason_example/BXAandBXDmice.RDS')
BXAandBXDleafvolumes <- readRDS('./jason_example/BXAandBXDleafvolumes.RDS')

#Create a copy of BXAtree
BXAandBXDtree <- Clone(BXAtree)

#Extract nodes as list
BXAandBXDnodes <- Traverse(BXAandBXDtree)
BXDtreenodes <- Traverse(BXDtree)

#Iterate over list
for (i in 1:length(BXAandBXDnodes)) {
  BXAandBXDnodes[[i]]$volumes <- c(BXAandBXDnodes[[i]]$volumes,
                                   BXDtreenodes[[i]]$volumes)
}
  
```

Not sure how traverse works but it seems to be able to influence the tree that it refers to. So this retroactively modifies `BXAandBXDtree`.

Write the combined data out

```{r}
# write_rds(BXAandBXDmice, "BXAandBXDmice.RDS")
# write_rds(BXAandBXDleafvolumes, "BXAandBXDleafvolumes.RDS")
```

Calculate heritability

```{r}
library(lme4)
calcH2 <- function(values, strains) {
  df <- data.frame(values=values, strains=strains)

  l <- lmer(values ~ 1  + (1|strains), df)
  
  vc <- as.data.frame(VarCorr(l))$vcov
  return(vc[1] / sum(vc))
}
```

```{r}
BXAandBXDtree$Do(function(x){
  x$relvolumes <- residuals(lm(x$volumes ~ BXAandBXDtree$volumes))
  x$H2abs <- calcH2(x$volumes, BXAandBXDmice$Strain)
  x$H2rel <- calcH2(x$relvolumes, BXAandBXDmice$Strain)
})
```

```{r}
library(grid)
library(MRIcrotome)
library(RMINC)

# nlin <- mincArray(mincGetVolume("/Users/jason/Dropbox/data/atlases/Dorr_2008_Steadman_2013_Ullmann_2013_Richards_2011_Qiu_2016_Egan_2015_40micron/ex-vivo/DSURQE_40micron_average.mnc"))
# labels <- mincArray(mincGetVolume("/Users/jason/Dropbox/data/atlases/Dorr_2008_Steadman_2013_Ullmann_2013_Richards_2011_Qiu_2016_Egan_2015_40micron/ex-vivo/DSURQE_40micron_labels.mnc"))

nlin <- mincArray(mincGetVolume('/projects/tools/atlases/Dorr_2008_Steadman_2013_Ullmann_2013_Richards_2011_Qiu_2016_Egan_2015_Elder_2019_40micron/ex-vivo/DSURQEE_40micron_average.mnc'))
labels <- mincArray(mincGetVolume('/projects/tools/atlases/Dorr_2008_Steadman_2013_Ullmann_2013_Richards_2011_Qiu_2016_Egan_2015_Elder_2019_40micron/ex-vivo/DSURQEE_40micron_labels.mnc'))
```


```{r}
#pdf("H2MRI.pdf", height=3, width=3, pointsize = 8)
pMRI <- sliceSeries(nrow=6, ncol=1, begin=75, end=350) %>%
  anatomy(nlin, low=700, high=1400) %>%
  addtitle("T2w MRI") %>%
  sliceSeries() %>%
  anatomy() %>%
  overlay(hanatToVolume(BXAandBXDtree, labels, "color_hex_triplet"), low=0, high=1) %>%
  addtitle("Atlas") %>%
  sliceSeries() %>%
  anatomy() %>%
  overlay(hanatToVolume(BXAandBXDtree, labels, "H2abs"), low=0.2, high=0.7) %>%
  addtitle("H2 (abs)") %>%
  sliceSeries() %>%
  anatomy() %>%
  overlay(hanatToVolume(BXAandBXDtree, labels, "H2rel"), low=0.2, high=0.7) %>%
  addtitle("H2 (rel)") %>%
  legend("H2") %>%
  draw()
#dev.off()
```

```{r}
sliceSeries(nrow = 6, ncol = 1, begin = 75, end = 350) %>% 
  anatomy(nlin, low = 700, high = 1400) %>% 
  sliceSeries() %>%
  anatomy() %>%
  overlay(hanatToVolume(BXAandBXDtree, labels, "H2abs"), low=0.2, high=0.7) %>%
  addtitle("H2 (abs)") %>%
  sliceSeries() %>%
  anatomy() %>%
  overlay(hanatToVolume(BXAandBXDtree, labels, "H2rel"), low=0.2, high=0.7) %>%
  addtitle("H2 (rel)") %>%
  legend("H2") %>%
  draw() 
```


Prepare for merging with hypothalamus HMDP RNA-seq data

```{r}
BXAandBXDmice <- BXAandBXDmice %>%
  mutate(Strain = fct_recode(
    Strain,
    "C57BL/6J" = "B6",
    "A/J" = "AJ",
    "DBA/2J" = "DBA",
    ))
```


```{r}
#absH2 <- apply(BXAandBXDleafvolumes, 2, function(x) calcH2(BXAandBXDmice, x))
#relH2 <- apply(BXAandBXDleafvolumes, 2, function(x) calcH2(BXAandBXDmice, x, type="rel"))
```

```{r}
#pdf("volsplot.pdf", height=3, width=3)
library(RColorBrewer)
tmpcols <- brewer.pal(3, "Set2")
annotframe <- data.frame(Strain=c("AXB-1", "BXD40"), 
                         value=rep(550, 2), 
                         name=rep("Brain", 2), 
                         label=c("BXA panel", "BXD panel"), 
                         Cohort=c("BXA", "BXD"))
BXAandBXDmice %>% 
  mutate(Thalamus=FindNode(BXAandBXDtree, "Thalamus")$relvolumes,
         HPC=FindNode(BXAandBXDtree, "Ammon's horn")$relvolumes,
         Brain=BXAandBXDtree$volumes) %>%
  pivot_longer(cols=c(Thalamus, HPC, Brain)) %>% 
  mutate(Strain = fct_inorder(Strain),
         Strain = fct_relevel(Strain, "C57BL/6J", after=6)) %>% 
  ggplot() + aes(x=Strain, y=value, colour=Cohort) + 
  #ggdist::stat_halfeye(adjust=1, width=.6, justification=-.2, .width=0) + 
  geom_boxplot(outlier.colour = NA) +
  #ggdist::stat_dots(side="left", justification=1.1, binwidth=.25) +
  ylab(bquote(Volume ~ (mm^3))) +
  #labs(title = "Volumes", caption="HPC and Thalamus residuals of brain volume regression") +
  facet_grid(name ~ ., scales = "free_y") + 
  geom_label(data=annotframe, aes(label=label), vjust=1) +
  #annotate("label", y=5, x=3, label="BXA", colour=tmpcols[1]) +
  #theme_gray(16) + 
  theme_bw(7) + 
  scale_color_brewer(palette = "Set2") +
  theme(axis.text.x = element_text(angle=45, hjust=1), 
        plot.background = element_blank(),
        legend.position = "none")
#dev.off()
```

Write data out in format readable by genenetwork

```{r}

BXDtree$Do(function(x){
  BXDmice$currentvol <- x$volumes
  x$relvolumes <- residuals(lm(currentvol ~ BV, BXDmice))
})

strainVols <- BXDmice %>%
  mutate(thalamus = FindNode(BXDtree, "Thalamus")$volumes,
         DCN = FindNode(BXDtree, "Cerebellar nuclei")$volumes,
         HPC = FindNode(BXDtree, "Ammon's horn")$volumes) %>%
  group_by(strain) %>%
  summarize(thalamus.sd = sd(thalamus),
            thalamus = mean(thalamus),
            DCN.sd = sd(DCN),
            DCN = mean(DCN),
            BV.sd = sd(BV),
            BV = mean(BV),
            HPC.sd=sd(HPC),
            HPC=mean(HPC))

strainVols  %>% mutate(strain = fct_recode(strain, "C57BL/6J" = "B6",
                                           "DBA/2J"="DBA"), HPC=HPC,
                       `@format=column` = strain) %>% 
  select(`@format=column`, HPC, BV, DCN) %>% 
  write_tsv("HPC-relvol.tsv")

```

Now, go here:

http://gn1.genenetwork.org/webqtl/main.py?FormID=batSubmit

Use these options:
1. BXD Family
2. the HPC-relvol.tsv file saved above
3. Some name (say HPC)

Click Next

On the page that opens then, click on the HPC trait (the hyperlink in the table)

Then, click on mapping tools

Click compute

How do I get the genes that show up? Not sure how to map loci to genes. 

Looks like I can zoom into a chromosome and a genetic locus. If I zoom in enough, there's a table on the page indicating which genes are in the region. 


Look up human genes here: https://open.win.ox.ac.uk/ukbiobank/big40/pheweb33k/


